### PR TITLE
620 bfl dataset card on dashboard highlights if dataset needs updating

### LINKED ISSUE #616: Brownfield land notification banner
## Context:

- LPAs are required to review their brownfield land data each year, and update their brownfield land register.
- They can use the Submit service to check and submit their data.
- The service should inform them that this needs to happen, and should guide them through the process of checking and submitting their data.
- This will take the form of a notification banner on the dashboard page.

## Rules for when the banner should show:

- The notification banner should only show when the LPA needs to update their brownfield land data.
- The long-term solution for this should be a data quality measure that comes via the pipeline (potentially either a set of rules that looks at when the latest endpoint was added or updated, or a more definitive label supplied by the user at time of submission), but in the interim we should bake the logic for this into the service.
- The [proposal for this is here](https://docs.google.com/document/d/1g8aKLlI8KAIHGaok-uvAmpELGx4pPpFdh0xpomkwRtM/edit?tab=t.0#heading=h.oi0xwdg862v).

## Acceptance criteria:

- As a data provider, when our brownfield land data needs to be updated, the service shows a notification on the dashboard page that alerts me to this, and links me to the step-by-step page to explain what I need to do.

## Designs:

Image

### PR DESCRIPTION
**Pull request by GeorgeGoodall-GovUk**
## What type of PR is this? (check all applicable)
- [ ] Feature

## Description
Adds notification banners for due and overdue datasets

## Related Tickets & Documents
- Closes #616 

## QA Instructions, Screenshots, Recordings
Preview Link: https://submit-pr-644.herokuapp.com/

## Added/updated tests?
_We encourage you to keep the code coverage percentage at 80% and above._
- Yes

## Release Notes

- **New Features**
  - Introduced deadline notices for datasets, enhancing user awareness of submission deadlines.
  - Added new HTML templates and macros for rendering deadline notifications.
  - Implemented middleware functions for managing dataset submission deadlines and notices.
  - Enhanced SQL queries to provide additional data fields related to deadlines.
  - Improved dataset overview and organisation overview pages with conditional rendering for notices.

- **Bug Fixes**
  - Improved error handling for dataset submission deadlines and notices.

- **Documentation**
  - Updated schemas to include new notice fields for datasets.

- **Tests**
  - Expanded test coverage for dataset notices and deadline handling functionalities, ensuring robust validation of new features.

### CODE REVIEW

File: src/views/components/deadlineNoticeContent.html [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+{% endmacro %}
+
+{% macro noticeHint(dataset) %}
+  {% if dataset.notice.type == 'overdue' %}
+    It was due on {{dataset.notice.deadline}}
+  {% endif %}
+{% endmacro %}
+
+{% macro noticeLink(organisation, slug) %}
+  <a class="govuk-notification-banner__link" href="organisations/{{organisation}}/{{slug}}/get-started">Follow the steps and check your data meets the specifications before you publish and submit it</a>.
```
Comment (alextea) [inline src/views/components/deadlineNoticeContent.html:16]: missing `/` at start of href

File: src/views/organisations/overview.html [hunk 2]
```diff
@@ -82,7 +94,9 @@ <h2 class="govuk-heading-m">
               {% endif %}
             </h2>
             <div class="govuk-task-list__hint">
-              {% if dataset.status == 'Not submitted' %}
+              {% if dataset.notice %}
+                {{ noticeHeading(dataset, dataset.notice) }}
```
Comment (alextea) [inline src/views/organisations/overview.html:98]: Content here should be `You must update this dataset by 31 December 2024` without the dataset name

File: src/middleware/datasetOverview.middleware.js [hunk 3]
```diff
@@ -90,20 +93,83 @@
...
+    let datasetSuppliedForCurrentYear = false
+    let datasetSuppliedForLastYear = false
+
+    const { deadlineDate, lastYearDeadline, twoYearsAgoDeadline } = getDeadlineHistory(deadlineObj.deadline)
+
+    const startDate = new Date(source.startDate)
+
+    datasetSuppliedForCurrentYear = startDate >= lastYearDeadline && startDate < deadlineDate
+    datasetSuppliedForLastYear = startDate >= twoYearsAgoDeadline && startDate < lastYearDeadline
+
```
[the remark this answers is not in the document]
Comment (GeorgeGoodall-GovUk) [reply src/middleware/datasetOverview.middleware.js]: @coderabbitai wont the new Date() throw an error if an invalid date is passed?

**Comment by CharliePatterson** [comment]
It looks like the check isn't working as intended on the dashboard page for LPAs who have already submitted BfL data this year - e.g. [Hull's dashboard](https://submit-pr-644.herokuapp.com/organisations/local-authority:KHL) is showing the notification banner and the text on the dataset details card.

Also, the banner isn't showing on [the BfL pages](https://submit-pr-644.herokuapp.com/organisations/local-authority:ADU/brownfield-land/overview).

**Review by rosado — DISMISSED** [review]

File: src/middleware/overview.middleware.js [hunk 4]
```diff
@@ -55,6 +55,110 @@ const fetchEntityCounts = async (req, res, next) => {
...
+      datasetSuppliedForLastYear = startDate >= twoYearsAgoDeadline && startDate < lastYearDeadline
+    } else {
+      resource = { dataset: dataset.dataset }
+    }
+
+    const warningDate = new Date(deadlineDate.getTime())
+    warningDate.setMonth(warningDate.getMonth() - dataset.noticePeriod)
+
+    const dueNotice = !datasetSuppliedForCurrentYear && currentDate > warningDate
+    const overdueNotice = !dueNotice && !datasetSuppliedForCurrentYear && !datasetSuppliedForLastYear
```
[the remark this answers is not in the document]
Comment (GeorgeGoodall-GovUk) [reply src/middleware/overview.middleware.js:108]: MJ and Charlie have said they want the due notice to take priority

**Comment by GeorgeGoodall-GovUk** [comment]
> BfL seems to be [showing up twice for Bristol](https://submit-pr-644.herokuapp.com/organisations/local-authority:BST)
> 
> Screenshot 2024-11-14 at 15 51 02 Screenshot 2024-11-14 at 15 51 45

should be good now

File: test/unit/views/organisations/dataset-overview.test.js [hunk 5]
```diff
@@ -116,6 +116,47 @@ describe('Dataset Overview Page', () => {
...
+
+    const htmlWithNotice = stripWhitespace(nunjucks.render('organisations/dataset-overview.html', paramsWithNotice))
+    const domWithNotice = new jsdom.JSDOM(htmlWithNotice)
+    const documentWithNotice = domWithNotice.window.document
+
+    expect(documentWithNotice.querySelector('.govuk-notification-banner')).not.toBeNull()
+    const banner = documentWithNotice.querySelector('.govuk-notification-banner')
+    expect(banner.classList.contains('govuk-notification-banner--warning')).toBeFalsy()
+    expect(banner.textContent).toContain(`You must update your ${paramsWithNotice.dataset.dataset} dataset by deadline`)
+  })
```
Comment (DilwoarH) [inline test/unit/views/organisations/dataset-overview.test.js:140]:
```suggestion
  it('Renders a notice when one is provided', () => {
    const paramsWithNotice = {
      ...params,
      notice: {
        type: 'due',
        deadline: 'deadline'
      }
    }

    const htmlWithNotice = stripWhitespace(nunjucks.render('organisations/dataset-overview.html', paramsWithNotice))
    const domWithNotice = new jsdom.JSDOM(htmlWithNotice)
    const documentWithNotice = domWithNotice.window.document
    const banner = documentWithNotice.querySelector('.govuk-notification-banner')

    expect(documentWithNotice.querySelector('.govuk-notification-banner')).not.toBeNull()
    expect(banner.classList.contains('govuk-notification-banner--warning')).toBeFalsy()
    expect(banner.textContent).toContain(`You must update your ${paramsWithNotice.dataset.dataset} dataset by deadline`)
  })
```

minor grouping change

File: test/unit/views/organisations/dataset-overview.test.js [hunk 6]
```diff
@@ -116,6 +116,47 @@ describe('Dataset Overview Page', () => {
...
+    }
+
+    const htmlWithNotice = stripWhitespace(nunjucks.render('organisations/dataset-overview.html', paramsWithNotice))
+    const domWithNotice = new jsdom.JSDOM(htmlWithNotice)
+    const documentWithNotice = domWithNotice.window.document
+
+    const banner = documentWithNotice.querySelector('.govuk-notification-banner')
+    expect(banner).not.toBeNull()
+    expect(banner.textContent).toContain(`Your ${paramsWithNotice.dataset.dataset} dataset is overdue`)
+  })
```
Comment (DilwoarH) [inline test/unit/views/organisations/dataset-overview.test.js:158]:
```suggestion
  it('Should render "overdue" notice with correct content', () => {
    const paramsWithNotice = {
      ...params,
      notice: {
        type: 'overdue',
        deadline: 'deadline'
      }
    }

    const htmlWithNotice = stripWhitespace(nunjucks.render('organisations/dataset-overview.html', paramsWithNotice))
    const domWithNotice = new jsdom.JSDOM(htmlWithNotice)
    const documentWithNotice = domWithNotice.window.document
    const banner = documentWithNotice.querySelector('.govuk-notification-banner')

    expect(banner).not.toBeNull()
    expect(banner.textContent).toContain(`Your ${paramsWithNotice.dataset.dataset} dataset is overdue`)
  })
```

File: test/unit/views/organisations/lpaOverviewPage.test.js [hunk 7]
```diff
@@ -127,17 +138,49 @@ describe(`LPA Overview Page (seed: ${seed})`, () => {
...
+
+        expect(banner.textContent).toContain(expectedHeader)
+        if (expectedHint) {
+          expect(banner.textContent).toContain(expectedHint)
+        }
+
+        const link = banner.querySelector('.govuk-notification-banner__link')
+
+        expect(link.getAttribute('href')).toEqual(expectedLinkHref)
+      })
```
Comment (DilwoarH) [inline test/unit/views/organisations/lpaOverviewPage.test.js:183]:
```suggestion
      it(`Renders the notice for dataset ${dataset.dataset}`, () => {
        let expectedHeader
        let expectedHint

        if (dataset.notice.type === 'due') {
          expectedHeader = `You must update your ${datasetSlugToReadableName(dataset.dataset).toLowerCase()} dataset by ${dataset.notice.deadline}`
        } else if (dataset.notice.type === 'overdue') {
          expectedHeader = `Your ${datasetSlugToReadableName(dataset.dataset).toLowerCase()} dataset is overdue`
          expectedHint = `It was due on ${dataset.notice.deadline}`
        }

        const expectedLinkHref = `/organisations/${params.organisation.organisation}/${dataset.dataset}/get-started`
        const link = banner.querySelector('.govuk-notification-banner__link')

        if (expectedHint) expect(banner.textContent).toContain(expectedHint)
        expect(banner.textContent).toContain(expectedHeader)
        expect(link.getAttribute('href')).toEqual(expectedLinkHref)
      })
```

**Review by DilwoarH — CHANGES_REQUESTED** [review]
Just some minor changes to test file structure. The most comment method is the Arrange-Act-Assert Pattern:

https://learn.microsoft.com/en-us/visualstudio/test/unit-test-basics?view=vs-2022#write-your-tests

```
it('should return the correct sum when adding two numbers', () => {
    // Arrange
    const a = 2, b = 3;

    // Act
    const result = add(a, b);

    // Assert
    expect(result).toBe(5);
});
```

**Comment by DilwoarH** [comment]
Code looks all good now - over to product and design sign off

@CharliePatterson @alextea

**Comment by alextea** [comment]
Looks good to me 🚀

**Review by DilwoarH — APPROVED** [review]

**Review by DilwoarH — APPROVED** [review]

**Pull request merged**